### ISSUE TITLE
Use setuid and setguid?

### ISSUE THREAD

**Issue by rohbotics** [UbiquityRobotics/pi_sonar#3]
Should we consider using this as a way to make running as root more feasible?

**Comment by jim-v** [comment]
The current build instructions involve a `sudo chmod 4755`, which makes the executable setuid. A problem with this is that the setuid bit gets reset every time `catkin_make` is run.

**Comment by rohbotics** [comment]
CMake can set the SETUID bit using the PERMISSIONS flag on the install target, this should solve this problem for running from debs.

**Comment by rohbotics** [comment]
Looks like the rpath isn't getting set in the installed version of the executable, so dynamic libraries cannot be resolved (you cannot use the `LD_LIBRARY_PATH` env variable with setuid). 

This command adds the needed rpath to the executable. 
`sudo patchelf --force-rpath --set-rpath '/opt/ros/kinetic/lib/' /opt/ros/kinetic/lib/pi_sonar/pi_sonar`

**Comment by jim-v** [comment]
`LD_LIBRARY_PATH` is considered harmful even without setuid.  We could create a file in `/etc/ld.so.conf.d/`

**Comment by rohbotics** [comment]
LD_LIBRARY_PATH is how ROS resolves dynamic libraries on installed packages usually.
It is part of setup.bash

**Comment by rohbotics** [comment]
Asked on ROS answers

https://answers.ros.org/question/293230/rpath-in-installed-executable/

**Comment by rohbotics** [comment]
Some further investigation:


This is the executable in catkin_ws/devel (readelf -d)
```
Dynamic section at offset 0x58ed0 contains 33 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libroscpp.so]
 0x00000001 (NEEDED)                     Shared library: [librosconsole.so]
 0x00000001 (NEEDED)                     Shared library: [libroscpp_serialization.so]
 0x00000001 (NEEDED)                     Shared library: [librostime.so]
 0x00000001 (NEEDED)                     Shared library: [libpthread.so.0]
 0x00000001 (NEEDED)                     Shared library: [libstdc++.so.6]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x00000001 (NEEDED)                     Shared library: [ld-linux-armhf.so.3]
 0x0000000f (RPATH)                      Library rpath: [/opt/ros/kinetic/lib:]
 0x0000000c (INIT)                       0x2ccc4
 0x0000000d (FINI)                       0x5dc98
 0x00000019 (INIT_ARRAY)                 0x78a40
 0x0000001b (INIT_ARRAYSZ)               8 (bytes)
 0x0000001a (FINI_ARRAY)                 0x78a48
 0x0000001c (FINI_ARRAYSZ)               4 (bytes)
 0x6ffffef5 (GNU_HASH)                   0x101b4
 0x00000005 (STRTAB)                     0x18318
 0x00000006 (SYMTAB)                     0x12b98
 0x0000000a (STRSZ)                      79507 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000015 (DEBUG)                      0x0
 0x00000003 (PLTGOT)                     0x79000
 0x00000002 (PLTRELSZ)                   1704 (bytes)
 0x00000014 (PLTREL)                     REL
 0x00000017 (JMPREL)                     0x2c61c
 0x00000011 (REL)                        0x2c5ac
 0x00000012 (RELSZ)                      112 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x6ffffffe (VERNEED)                    0x2c49c
 0x6fffffff (VERNEEDNUM)                 5
 0x6ffffff0 (VERSYM)                     0x2b9ac
 0x00000000 (NULL)                       0x0
```

This is the executable in catkin_ws/install
```
Dynamic section at offset 0x58ed0 contains 32 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libroscpp.so]
 0x00000001 (NEEDED)                     Shared library: [librosconsole.so]
 0x00000001 (NEEDED)                     Shared library: [libroscpp_serialization.so]
 0x00000001 (NEEDED)                     Shared library: [librostime.so]
 0x00000001 (NEEDED)                     Shared library: [libpthread.so.0]
 0x00000001 (NEEDED)                     Shared library: [libstdc++.so.6]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so.6]
 0x00000001 (NEEDED)                     Shared library: [ld-linux-armhf.so.3]
 0x0000000c (INIT)                       0x2ccc4
 0x0000000d (FINI)                       0x5dc98
 0x00000019 (INIT_ARRAY)                 0x78a40
 0x0000001b (INIT_ARRAYSZ)               8 (bytes)
 0x0000001a (FINI_ARRAY)                 0x78a48
 0x0000001c (FINI_ARRAYSZ)               4 (bytes)
 0x6ffffef5 (GNU_HASH)                   0x101b4
 0x00000005 (STRTAB)                     0x18318
 0x00000006 (SYMTAB)                     0x12b98
 0x0000000a (STRSZ)                      79507 (bytes)
 0x0000000b (SYMENT)                     16 (bytes)
 0x00000015 (DEBUG)                      0x0
 0x00000003 (PLTGOT)                     0x79000
 0x00000002 (PLTRELSZ)                   1704 (bytes)
 0x00000014 (PLTREL)                     REL
 0x00000017 (JMPREL)                     0x2c61c
 0x00000011 (REL)                        0x2c5ac
 0x00000012 (RELSZ)                      112 (bytes)
 0x00000013 (RELENT)                     8 (bytes)
 0x6ffffffe (VERNEED)                    0x2c49c
 0x6fffffff (VERNEEDNUM)                 5
 0x6ffffff0 (VERSYM)                     0x2b9ac
 0x00000000 (NULL)                       0x0
```

Looks like something in the cmake install process is removing the RPATH from the executable.

**Comment by rohbotics** [comment]
Adding `set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)` to the CMake solves the issue.

**Comment by rohbotics** [comment]
Fixed in 0.3.1, deb available here https://packages.ubiquityrobotics.com/ubuntu/ubiquity-testing/pool/main/r/ros-kinetic-pi-sonar/ros-kinetic-pi-sonar_0.3.1-0xenial~tbpo1_armhf.deb
